solvation_analysis/analysis_library.py

```diff
@@ -0,0 +1,227 @@
+            a specification for a shell composition. Keys are residue names (str)
+            and values are the number of desired residues. e.g. if shell_dict =
+            {'mol1': 4} then the function will return the percentage of shells
+            that have 4 mol1. Note that this may include shells with 4 mol1 and
+            any number of other solvents. To specify a shell with 4 mol1 and nothing
+            else, enter a dict such as {'mol1': 4, 'mol2': 0, 'mol3': 0}.
+
+        Returns
+        -------
+        float
+            the percentage of shells
+        """
+        query_list = [f"{name} == {str(count)}" for name, count in shell_dict.items()]
+        query = " and ".join(query_list)
+        query_counts = self.speciation_percent.query(query)
+        return query_counts['count'].sum()
+
+    def find_shells(self, shell_dict):
+        """
+        Find all solvation shells that match shell_dict.
+
+        This returns the frame, solute index, and composition of all solutes
+        that match the composition given in shell_dict.
+
+        Attributes
+        ----------
+        shell_dict : dict of {str: int}
+            a specification for a shell composition. Keys are residue names (str)
+            and values are the number of desired residues. e.g. if shell_dict =
+            {'mol1': 4} then the function will return all shells
+            that have 4 mol1. Note that this may include shells with 4 mol1 and
+            any number of other solvents. To specify a shell with 4 mol1 and nothing
+            else, enter a dict such as {'mol1': 4, 'mol2': 0, 'mol3': 0}.
+
+        Returns
+        -------
+        pandas.DataFrame
+            the index and composition of all shells that match shell_dict
+        """
+        query_list = [f"{name} == {str(count)}" for name, count in shell_dict.items()]
+        query = " and ".join(query_list)
+        query_counts = self.speciation.query(query)
+        return query_counts
+
+
+class Coordination:
+    """
+    Calculate the coordination number of each solvent.
+
+    Coordination calculates the coordination number by averaging the number of
+    coordinated solvents in all of the solvation shells. This is equivalent to
+    the typical method of integrating the RDF up to the solvation radius cutoff.
+
+    The coordination numbers are made available as an average over the whole
+    simulation and by frame.
+
+    Parameters
+    ----------
+    solvation_data : pandas.DataFrame
+        The solvation data frame output by Solution.
+    n_frames : int
+        The number of frames in solvation_data.
+    n_solutes : int
+        The number of solutes in solvation_data.
+
+    Attributes
+    ----------
+    cn_dict : dict
+        a dictionary where keys are residue names (str) and values are the
+        average coordination number of that residue (float).
+    cn_by_frame : pd.DataFrame
+        a dictionary tracking the average coordination number of each
```

Comment:
NIT: This is a dataframe not a dict?

Reply:
This is a dict, just for slightly easier selection